wrapper/src/main/java/software/amazon/jdbc/plugin/cache/CacheConnection.java

```diff
@@ -294,33 +348,27 @@ private void initializeCacheConnectionIfNeeded(boolean isRead) {
       }
     }
 
-    // Register cluster with CacheMonitor on first cache operation (skip if telemetryFactory is null = test mode)
-    if (telemetryFactory != null && !cacheMonitorRegistered) {
-      synchronized (this) {
-        if (!cacheMonitorRegistered) {
-          CacheMonitor.registerCluster(
-              inFlightWriteSizeLimitBytes, healthCheckInHealthyState, telemetryFactory,
-              this.cacheRwServerAddr, this.cacheRoServerAddr,
-              this.useSSL, this.cacheConnectionTimeout, this.iamAuthEnabled, this.credentialsProvider,
-              this.cacheIamRegion, this.cacheName, this.cacheUsername, this.cachePassword
-          );
-          cacheMonitorRegistered = true;
-        }
+    connectionInitializationLock.lock();
+    try {
+      // Detect cluster mode first (cached for reuse)
+      detectClusterMode();
+
+      // Register cluster with CacheMonitor on first cache operation (skip if telemetryFactory is null = test mode)
+      if (telemetryFactory != null && !cacheMonitorRegistered) {
+        CacheMonitor.registerCluster(
+            inFlightWriteSizeLimitBytes, healthCheckInHealthyState, telemetryFactory,
+            this.cacheRwServerAddr, this.cacheRoServerAddr,
+            this.useSSL, this.cacheConnectionTimeout, this.iamAuthEnabled, this.credentialsProvider,
+            this.cacheIamRegion, this.cacheName, this.cacheUsername, this.cachePassword
+        );
+        cacheMonitorRegistered = true;
       }
-    }
 
-    GenericObjectPool<StatefulRedisConnection<byte[], byte[]>> cacheConnectionPool =
-        isRead ? readConnectionPool : writeConnectionPool;
-    if (cacheConnectionPool == null) {
-      ReentrantLock connectionPoolLock = isRead ? READ_LOCK : WRITE_LOCK;
-      connectionPoolLock.lock();
-      try {
-        if ((isRead && readConnectionPool == null) || (!isRead && writeConnectionPool == null)) {
+      if ((isRead && readConnectionPool == null) || (!isRead && writeConnectionPool == null)) {
```

Comment:
I would move this logic to the start of this method. i.e. if the connection pool you are looking for is already initialized, then simply return.